server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1182,17 +1182,17 @@ private boolean isEmptyOrTrivialWildcard(List<String> expressions) {
 
         private static List<String> resolveEmptyOrTrivialWildcard(IndicesOptions options, Metadata metadata) {
             if (options.expandWildcardsOpen() && options.expandWildcardsClosed() && options.expandWildcardsHidden()) {
-                return Arrays.asList(metadata.getConcreteAllIndices());
+                return List.copyOf(metadata.getConcreteAllIndices());
```

Comment:
This would be expensive operation to copy the set to a list every time for index resolution.
Have run micro benchmark on your changes?

Reply:
Here ConcreteAllIndices(allIndices) are unmodifiableSet, so List.copyOf() will not really create a copy . 
There is an implementation Note in javadoc of this API - "If the given Collection is an unmodifiable List, calling copyOf will generally not create a copy". Also if we look in to the internal implementation of Lisy.copyOf() it's clear that, it's not creating a copy if collection is an AbstractImmutableList.

Reply:
you are doing
```Collections.unmodifiableSet(allIndices);``` which creates new object of ```UnmodifiableSet```. ```UnmodifiableSet``` extends ```UnmodifiableCollection``` and it is not extending AbstractImmutableList. It will end copying the elements. Can you cross check?

Reply:
Yes, you are correct. So what do you suggest here?
concrete indices treated as Set in the existing implementation - https://github.com/opensearch-project/OpenSearch/blob/67a2e4c7275afa93ce2c6fc2107ca0f7a8c461bd/server/src/main/java/org/opensearch/cluster/metadata/Metadata.java#L1636
That's why I kept only concrete indices as Set and all other indices as List. If we keep concrete indices also as List, then this conversion won't be required.

Reply:
yes right, lets check the usage in the code if it has to be Set, else i agree keeping it as list would be better.

Reply:
From the usage wise, it kept as Set to avoid duplicates in "indices" and to have better performance during "retainAll()". 
From the java doc of retainAll() - "If the specified collection is also a set, this operation effectively modifies this set so that its value is the intersection of the two sets."

So I don't think we can use List instead of Set here.

Reply:
How do we know there is not a regression here, given that `Arrays.asList` was previously a lightweight operation but this is now a per-element copy?

Reply:
@sandeshkr419 @andrross @shwetathareja Please review

Reply:
> I performed the performance testing as @sandeshkr419 suggested below. I created an index and 50000 aliases initially, then again created 500*10 = 5000 aliases (so total 55000) and calculated the average time taken for these last 500 sets of alias creation separately. The results shows that there is a consistent improvement in each sets(~10ms) after this change.


Thank @akolarkunnu for performing the perf runs. Since there is a change in ClusterStateHealth, we might need a similar run for `health` API calls. Could we also capture the memory allocations corresponding to the duration of run and compare them before / after change ? 

Also, trying to understand as to why 50K aliases were created and not the actual index ?

Reply:
> Also, trying to understand as to why 50K aliases were created and not the actual index ?

We could also create indices, its just you need to modify node limits on max indices/shards per node - plus the the change is cluster state is identical (as in same/similar objects are modified). 


> Since there is a change in ClusterStateHealth, we might need a similar run for health API calls. Could we also capture the memory allocations corresponding to the duration of run and compare them before / after change ?

I don't think we would be able to capture changes (improvements actually) here, the only thing we have omitted here is back and worth list-set conversions for some objects (stored as one type, retrieved as different). [Recalling from my own PR when I had tried to benchmark some changes in cluster state]

Reply:
Health API iterates over the IndexMetadata to compute the overall cluster health. Hence i was thinking that creating indexes rather than alias would have more entries to iterate for health computation. For cluster health computation, there were less memory allocations seen when it was switched to size based array iteration. Ref PR :- https://github.com/opensearch-project/OpenSearch/pull/15492

Please take a look at the memory allocations in flame graph in the referenced PR. Could we verify that switching from array to set iterator does not incur additional memory allocations for health API.

Reply:
@akolarkunnu Can you please take these changes to closure by profiling health API as in https://github.com/opensearch-project/OpenSearch/pull/15492 

That should bring confidence on whether or not to get these changes in code or not.

Reply:
Ok, I will resume this work soon.